### PR TITLE
feature: turn dummy module into simple example module

### LINKED ISSUE #2359: Turn dummy module into simple example module
Currently we don't get a lot of value out of our dummy module, but still need to modify it during refactors.  Instead of stubbing out implementations we could create a simple example module for:

- An ideal starting point for module writers that demonstrates every basic operation once - an API call, a transaction, a consensus item
- Keeping track of the boiler-plate code that we should try to eliminate for module writers
- Enable a "money printer" module that will make it easier to run integration tests
- Getting 90% of the way towards a proper account-based module (requested by some orgs)

This is also a necessary step towards #2315

### A useful example should include:
**Server**
- Config gen with a private and consensus key
- Example of a consensus item and threshold signing
- DB migrations
- Validation of input/outputs amounts and keys
- Audit of assets/liabilities
- API endpoint that notifies consensus
- API endpoint that blocks on db

**Client**
- Primary module impl
- Create a transaction between users
- Parse an output outcome
- Server API calls
- Simple state machine for refunding tx input and claiming tx output

**Tests**
- A user-to-user transaction
- API calls
- Refund / output claiming
- Submitting invalid txs

@elsirion @justinmoon ^ any things I'm missing as far as a useful example is concerned?

### PR DESCRIPTION
**Pull request by jkitman**
Fixes #2359 (see there for details)

### CODE REVIEW

File: modules/fedimint-dummy-tests/tests/tests.rs
Comment (justinmoon) [inline modules/fedimint-dummy-tests/tests/tests.rs:1]: Does this mean that the fixtures now work with external modules to some extent?

Comment (jkitman) [reply modules/fedimint-dummy-tests/tests/tests.rs:1]: Yep this is the first passing test written in the modular test framework :smile: 

Over the next couple weeks I'll be converting our old tests to use the new framework to test the new clients.

Comment (justinmoon) [reply modules/fedimint-dummy-tests/tests/tests.rs:1]: Amazing

File: modules/fedimint-dummy-client/src/lib.rs
Comment (m1sterc001guy) [inline modules/fedimint-dummy-client/src/lib.rs:87]: Do you want to commit the transaction here? Wouldn't this deduct the balance before the transaction is accepted by the fed?

Not sure how robust we want to make the dummy module, but I think one way you could do it would be to spawn a state machine that waits on the output of this TX and then updates the local balance after verifying that the transaction was successful.

Comment (m1sterc001guy) [reply modules/fedimint-dummy-client/src/lib.rs:87]: Or maybe just move `commit_tx` underneath `await_tx_accepted`

File: modules/fedimint-dummy-client/src/states.rs
Comment (m1sterc001guy) [inline modules/fedimint-dummy-client/src/states.rs:29]: Why do we need both of these? They look the same to me

File: modules/fedimint-dummy-client/src/states.rs
Comment (m1sterc001guy) [inline modules/fedimint-dummy-client/src/states.rs:72]: Is the `clone` necessary?

File: modules/fedimint-dummy-common/src/db.rs
Comment (m1sterc001guy) [inline modules/fedimint-dummy-common/src/db.rs:14]: Nit: not a big deal for the dummy module but I think its a bit weird to have a shared db.rs for the client/server since I don't think they shouldn't ever access the same database structures

File: modules/fedimint-dummy-common/src/db.rs
Comment (m1sterc001guy) [inline modules/fedimint-dummy-common/src/db.rs:81]: Does `DummyFundsKeyV1` have the same data as `DummyFundsKeyV0`? Can we make them different so that the migration actually changes the data structure?

File: modules/fedimint-dummy-common/src/lib.rs
Comment (m1sterc001guy) [inline modules/fedimint-dummy-common/src/lib.rs:47]: Copy error?

File: modules/fedimint-dummy-common/src/lib.rs
Comment (m1sterc001guy) [inline modules/fedimint-dummy-common/src/lib.rs:55]: Copy error?

**Comment by m1sterc001guy** [comment]
Really nice PR! I'll let others take a look before merging since there's a lot of new code added. Most of my comments are nits

File: modules/fedimint-dummy-client/src/api.rs
Comment (jkitman) [reply modules/fedimint-dummy-client/src/api.rs:9]: Maybe `Example`?  It's 90% of the way to being a proper `Account` based module as well.  I'll defer this to another PR.

File: modules/fedimint-dummy-client/src/lib.rs
Comment (jkitman) [reply modules/fedimint-dummy-client/src/lib.rs:87]: Does `finalize_and_submit_transaction` start the input state machine?  If so, there is already a refund in that state machine.

The state machine is the trickest part of understand.

File: modules/fedimint-dummy-client/src/states.rs
Comment (jkitman) [reply modules/fedimint-dummy-client/src/states.rs:29]: It's so the client can track their balance independently of the fed (similar to how the mint module works).

If input fails, we get refunded.  If output fails, we don't.
If input succeeds, nothing happens since we already deducted.  If output succeeds, we add new funds.

File: modules/fedimint-dummy-common/src/db.rs
Comment (jkitman) [reply modules/fedimint-dummy-common/src/db.rs:14]: Was following a pattern in the other modules, but I think you are right.

File: modules/fedimint-dummy-common/src/db.rs
Comment (jkitman) [reply modules/fedimint-dummy-common/src/db.rs:81]: Sure, it's mostly for an example anyway.

File: modules/fedimint-dummy-client/src/lib.rs
Comment (jkitman) [reply modules/fedimint-dummy-client/src/lib.rs:87]: I will add tests for these later.

File: modules/fedimint-dummy-client/src/lib.rs
Comment (elsirion) [inline modules/fedimint-dummy-client/src/lib.rs:86]: Why are we making the assumption that the user wants to spend from the dummy module wallet? Idiomatically one would create an unfunded transaction here and leave it to `finalize_and_submit_transaction` to fund it using the primary module. That *might* be the dummy module if configured, but not necessarily.

File: modules/fedimint-dummy-server/src/lib.rs
Comment (elsirion) [inline modules/fedimint-dummy-server/src/lib.rs:397]: Why not just use a transaction with a special-cased input/output for that? That might also solve #2387

File: modules/fedimint-dummy-client/src/lib.rs
Comment (elsirion) [inline modules/fedimint-dummy-client/src/lib.rs:171]: I wonder how much of this could become derive macros …

File: modules/fedimint-dummy-common/src/lib.rs
Comment (elsirion) [inline modules/fedimint-dummy-common/src/lib.rs:48]: I think it would be prudent to include some sort of nonce, otherwise a two separate transfers between the same accounts of the same value would be identical (same txid) and thus the second would not go through. This won't happen for a while, but when it does it will be hard to debug.

**Review by elsirion — APPROVED** [review]
I'm ok with it's current form, some nits and things that could be improved imo, but can happen as we go.

File: modules/fedimint-dummy-client/src/lib.rs
Comment (jkitman) [reply modules/fedimint-dummy-client/src/lib.rs:86]: :+1: That could be useful for funding our ecash wallet...

File: modules/fedimint-dummy-server/src/lib.rs
Comment (jkitman) [reply modules/fedimint-dummy-server/src/lib.rs:397]: Thought about it, but I wanted an example of how to use consensus items and threshold signatures instead of another transaction example.  Maybe I could demo some other way?

**Pull request merged**
